crates/floresta-chain/src/pruned_utreexo/consensus.rs

```diff
@@ -233,14 +380,51 @@ impl Consensus {
         };
         Ok(())
     }
-    #[allow(unused)]
+    /// Validate Locktime. if mtp is 0, skip the validation.
     fn validate_locktime(
         input: &TxIn,
         transaction: &Transaction,
+        out_map: &UtxoMap,
         height: u32,
+        mtp: u32,
     ) -> Result<(), BlockValidationErrors> {
-        unimplemented!("validate_locktime")
+        if mtp == 0 {
+            return Ok(());
+        }
+        if input.sequence.is_relative_lock_time() {
+            //validate lock time
+            let prevout = out_map.get(&input.previous_output).unwrap();
+            if input.sequence.is_height_locked() {
+                let commited_height = prevout.get_height();
+                //to retrieve the span contained in the sequence we have to just get the u32 value that the sequence contains.
+                let height_span = input.sequence.0;
+
+                // if the commited height + the span is greater than the current height, the transaction is invalid.
+                if commited_height + height_span > height {
+                    return Err(BlockValidationErrors::BadRelativeLockTime);
+                }
+            }
+            if input.sequence.is_time_locked() {
+                let commited_time = prevout.get_time();
+                // here we have to shift the sequence 16 bits to the left and 16 to the right get the time lock without the flag messing with us.
+                let time_lock = ((input.sequence.0 << 16) >> 16) * 512 as u32;
```

Comment:
I think shifting left and right to zero the MSB is weird and unsound. If you want the LSB, just & with 0xffff.

Reply:
Nice idea.